File: python/primitives.py
"""
Primitive Types in Python
Demonstrates all basic primitive types and their usage
"""

def main():
    print("=== Python Primitive Types ===\n")

    # Integer types (Python 3 has arbitrary precision integers)
    print("--- Integer Types ---")
    small_int = 42
    large_int = 123456789012345678901234567890
    negative_int = -999
    hex_int = 0xFF  # 255 in hexadecimal
    octal_int = 0o77  # 63 in octal
    binary_int = 0b1010  # 10 in binary

    print(f"Small integer: {small_int}")
    print(f"Large integer: {large_int}")
    print(f"Negative: {negative_int}")
    print(f"Hexadecimal (0xFF): {hex_int}")
    print(f"Octal (0o77): {octal_int}")
    print(f"Binary (0b1010): {binary_int}")

    # Floating point types
    print("\n--- Floating Point Types ---")
    float_num = 3.14159
    scientific = 1.5e10  # 15000000000.0
    negative_float = -2.718

    print(f"Float: {float_num}")
    print(f"Scientific notation: {scientific}")
    print(f"Negative float: {negative_float}")

    # Complex numbers
    print("\n--- Complex Numbers ---")
    complex_num = 3 + 4j
    another_complex = complex(2, -5)

    print(f"Complex number: {complex_num}")
    print(f"Another complex: {another_complex}")
    print(f"Real part: {complex_num.real}")
    print(f"Imaginary part: {complex_num.imag}")

    # Boolean type
    print("\n--- Boolean Type ---")
    is_true = True
    is_false = False

    print(f"True: {is_true}")
    print(f"False: {is_false}")
    print(f"Boolean from comparison: {5 > 3}")

    # String types
    print("\n--- String Types ---")
    single_quote = 'Hello'
    double_quote = "World"
    multi_line = """This is a
    multi-line
    string"""
    raw_string = r"C:\path\to\file"
    f_string = f"Interpolated: {small_int}"

    print(f"Single quote: {single_quote}")
    print(f"Double quote: {double_quote}")
    print(f"Multi-line: {multi_line}")
    print(f"Raw string: {raw_string}")
    print(f"F-string: {f_string}")

    # Bytes type
    print("\n--- Bytes Type ---")
    byte_literal = b"Hello"
    byte_array = bytearray(b"World")

    print(f"Bytes: {byte_literal}")
    print(f"Bytearray: {byte_array}")

    # None type
    print("\n--- None Type ---")
    none_value = None
    print(f"None: {none_value}")
    print(f"Type: {type(none_value)}")

    # List (mutable sequence)
    print("\n--- List Type ---")
    numbers = [1, 2, 3, 4, 5]
    mixed = [1, "two", 3.0, True]

    print(f"Numbers list: {numbers}")
    print(f"Mixed list: {mixed}")
    print(f"First element: {numbers[0]}")

    # Tuple (immutable sequence)
    print("\n--- Tuple Type ---")
    coordinates = (10, 20)
    single_item = (42,)  # Note the comma
    mixed_tuple = (1, "hello", 3.14)

    print(f"Tuple: {coordinates}")
    print(f"Single item tuple: {single_item}")
    print(f"Mixed tuple: {mixed_tuple}")

    # Set (unordered unique elements)
    print("\n--- Set Type ---")
    unique_numbers = {1, 2, 3, 4, 5}
    unique_numbers.add(3)  # Won't add duplicate

    print(f"Set: {unique_numbers}")
    print(f"Length: {len(unique_numbers)}")

    # Dictionary (key-value pairs)
    print("\n--- Dictionary Type ---")
    person = {
        "name": "Alice",
        "age": 30,
        "city": "New York"
    }

    print(f"Dictionary: {person}")
    print(f"Name: {person['name']}")

    # Type checking
    print("\n--- Type Checking ---")
    print(f"Type of integer: {type(small_int)}")
    print(f"Type of float: {type(float_num)}")
    print(f"Type of string: {type(single_quote)}")
    print(f"Type of list: {type(numbers)}")
    print(f"Type of dict: {type(person)}")

if __name__ == "__main__":
    main()

File: rust/primitives.rs
// Primitive Types in Rust
// Demonstrates all basic primitive types and their usage

fn main() {
    println!("=== Rust Primitive Types ===\n");

    // Integer types
    println!("--- Integer Types ---");
    let signed_8: i8 = -128;
    let unsigned_8: u8 = 255;
    let signed_16: i16 = -32768;
    let unsigned_16: u16 = 65535;
    let signed_32: i32 = -2147483648;
    let unsigned_32: u32 = 4294967295;
    let signed_64: i64 = -9223372036854775808;
    let unsigned_64: u64 = 18446744073709551615;
    let signed_128: i128 = -170141183460469231731687303715884105728;
    let unsigned_128: u128 = 340282366920938463463374607431768211455;
    let arch_signed: isize = -2147483648; // size depends on architecture
    let arch_unsigned: usize = 4294967295;

    println!("i8: {}", signed_8);
    println!("u8: {}", unsigned_8);
    println!("i32: {}", signed_32);
    println!("u64: {}", unsigned_64);
    println!("usize: {}", arch_unsigned);

    // Floating point types
    println!("\n--- Floating Point Types ---");
    let float_32: f32 = 3.14159;
    let float_64: f64 = 2.718281828459045;
    println!("f32: {}", float_32);
    println!("f64: {}", float_64);

    // Boolean type
    println!("\n--- Boolean Type ---");
    let is_true: bool = true;
    let is_false: bool = false;
    println!("true: {}", is_true);
    println!("false: {}", is_false);

    // Character type (Unicode scalar value)
    println!("\n--- Character Type ---");
    let letter: char = 'A';
    let emoji: char = '🦀';
    let chinese: char = '中';
    println!("ASCII char: {}", letter);
    println!("Emoji: {}", emoji);
    println!("Unicode: {}", chinese);

    // String types
    println!("\n--- String Types ---");
    let string_slice: &str = "Hello, Rust!"; // string slice (immutable)
    let owned_string: String = String::from("Hello, World!"); // owned string
    println!("String slice: {}", string_slice);
    println!("Owned String: {}", owned_string);

    // Array (fixed size)
    println!("\n--- Array Type ---");
    let numbers: [i32; 5] = [1, 2, 3, 4, 5];
    println!("Array: {:?}", numbers);
    println!("First element: {}", numbers[0]);

    // Tuple (fixed size, mixed types)
    println!("\n--- Tuple Type ---");
    let tuple: (i32, f64, char) = (42, 3.14, 'x');
    println!("Tuple: {:?}", tuple);
    println!("First: {}, Second: {}, Third: {}", tuple.0, tuple.1, tuple.2);

    // Unit type (empty tuple)
    println!("\n--- Unit Type ---");
    let unit: () = ();
    println!("Unit type (empty): {:?}", unit);
}
